apps/keyboard/js/imes/latin/worker.js

```diff
@@ -28,111 +31,282 @@
 //    first to quickly tell the keyboard that the word is valid.
 //
 
-'use strict';
-
 // Load the predictions.js module.  This defines the Predictions object.
 importScripts('predictions.js');
 
-// When we receive messages, translate them into function calls to one
-// of the functions in the Commands object below.
-self.onmessage = function(e) {
-  Commands[e.data.cmd].apply(null, e.data.args);
+// We invoke a "null" predictor when user dictionary is empty (incoming dict
+// blob is "undefined"). This is to avoid modifying the complex predictions.js
+// to adapt to "undefined" dict blob.
+
+// The switching mechanism is implemented in ProxyUserDictPredictor depending on
+// its setDictionary parameter when called. Note that PUDP.setNearByKeys is
+// always called upon userDictPredictor for setting the predictor up.
+
+// For pattern conformity in this file we don't use "new" to instantiate the
+// null and proxy instances.
+
+var NullPredictions = function(){
+  return {
+    predict: function(input, maxSuggestions, maxCandidates, maxCorrections,
+               callback, onerror) {
+      var status = {
+        state: 'predicting',
+        abort: function nullAbort() {
+          if (this.state !== 'done' && this.state !== 'aborted') {
+            this.state = 'aborting';
+          }
+        }
+      };
+
+      setTimeout(function nullFinish() {
+        if (status.state === 'aborting') {
+          status.state = 'aborted';
+        }
+
+        if (status.state === 'aborted') {
+          return;
+        }
+
+        status.state = 'done';
+        callback([]);
+      });
+
+      return status;
+    }
+  };
 };
 
-// Send console messages back to the main thread with this method
-function log(msg) {
-  postMessage({cmd: 'log', message: msg});
-}
+var ProxyUserDictPredictor = function(){
+  var userDictPredictor = Predictions();
+  var nullPredictor = NullPredictions();
+
+  var activePredictor;
+
+  return {
+    setDictionary: function(buffer) {
+      if (buffer === undefined) {
+        activePredictor = nullPredictor;
+      } else {
+        activePredictor = userDictPredictor;
+        userDictPredictor.setDictionary(buffer);
+      }
+    },
+    setNearbyKeys: function(data) {
+      userDictPredictor.setNearbyKeys(data);
+    },
+    // Directly pass arguments to wrapped predictor
+    predict: function(...args) {
```

Comment:
I can definitely ES5-fy it, but I guess those arrow functions need ES5-fication too?